src/Microsoft.DocAsCode.Build.Engine/DocumentBuildContext.cs

```diff
@@ -193,6 +200,111 @@ private List<string> ResolveByExternalReferencePackages(List<string> uidList, Co
             return list;
         }
 
+        private async Task<List<string>> ResolveByXRefServiceAsync(List<string> uidList, ConcurrentDictionary<string, XRefSpec> externalXRefSpec)
+        {
+            if (XRefServiceUrls == null || XRefServiceUrls.Length == 0)
+            {
+                return uidList;
+            }
+            int needQueryCount = uidList.Count;
+            var unresolvedUidList = new List<string>();
+            int pieceSize = 1000;
+            foreach (string requestUrl in XRefServiceUrls)
+            {    
+                for(int i = 0; i < uidList.Count; i+= pieceSize)
+                {
+                    List<string> smallPiece = uidList.GetRange(i, Math.Min(pieceSize, uidList.Count - i));
+                    var result = await QueryByHttpRequestAsync(requestUrl, smallPiece);
+                    bool isSuccess = result.Item1;
+                    List<XRefSpec> xsList = result.Item2;
+                    if (isSuccess)
+                    {
+                        for (int j = 0; j < xsList.Count; j++)
```

Comment:
it is sure that count here match smallPiece.Count because result responding to each uid is on the same index position, if not resolved then null will returned.

---
In reply to: [127369568](https://github.com/dotnet/docfx/pull/1837#discussion_r127369568) [](ancestors = 127369568)